big-o/big-o-analysis.md

```diff
@@ -0,0 +1,335 @@
+
+   1. `if item == array[i]` can happen _n_ times
+   1. `i in range(len(array))` can happen _n_ times
+      - If this is not obvious, this is a detail that we can leave as an unknown. We may have other clues in the rest of the process that could help us. We can revisit this if we realize that this is crucial to our solution.
+
+4. Create a formula that counts how many operations there are:
+
+   1. Two operations can happen _n_ times. We can express this as _2 \* n_, or _2n_
+
+5. Drop the constants:
+
+   1. In math, when looking at _2n_, `2` is a constant that we can drop
+   1. After dropping the constant, our remaining formula is _n_
+
+6. Match this formula with the most relevant complexity:
+   1. _n_ matches _O(n)_ the most closely. This algorithm has linear time complexity.
+
+### Large Example
+
+This is one implementation of reversing the order of items in a list. This should affect the original list.
+
+It loops through an array, and starts looking at the first item (at index `i`) and the last item (at index `j`), and swaps their values, making use of a temporary variable `temp`. It will return the list in reverse order.
+
+```python
+def reverse(array):
+    if len(array) <= 1:
+        return array
+
+    i = 0
+    j = len(array) - 1
+
+    while i < j:
+        temp = array[i]
+        array[i] = array[j]
+        array[j] = temp
+
+        i += 1
+        j -= 1
+
+    return array
+```
+
+1. Identify all lists with variable size: There is one list with variable size, and it's a size of _n_.
+
+2. Identify all operations in this algorithm:
+
+   1. `len(array) <= 1`
+   1. `i = 0`
+   1. `j = ...`
+   1. `len(array) - 1`
+   1. `i < j`
+   1. `temp = ...`
+   1. `array[i] = ...`
+   1. `array[j] = ...`
+   1. `i += 1`
+   1. `j -= 1`
+
+3. Recognize which operations have a relationship to list size:
+
+   1. All 4 of the operations outside of the `while` loop will only happen once; they have no relationship to list size
+   1. All 6 of the operations inside the `while` loop can happen _n_ times
+
+4. Create a formula that counts how many operations there are:
+
+   1. We can express this as _4 + (6 \* n)_, or _4 + 6n_
+
+5. Drop the constants:
+
+   1. We are able to drop `4 +` and `6 *`
+   1. After dropping the constants, our remaining formula is _n_
+
+6. Match this formula with the most relevant complexity:
+   1. _n_ matches _O(n)_ the most closely. This algorithm has linear time complexity.
+
+### What is "Dropping the Constant," or Finding the Dominant Order?
+
+In complexity analysis, programmers care about the dominant order, or the part of the complexity that affect the complexity _the most._
+
+Mathematical constants are values that have a fixed value. Constants tend to be added, subtracted, multiplied, or divided in our formulas. In Big O, constants will affect our complexity (O(6n) is different than O(n))... but they do not affect the curve enough that is meaningful for complexity analysis.
+
```

Comment:
Constants are emphasized, but this also good place to mention that all terms except the largest are dropped.